bluetooth: ANS: Add Alert Notification Service

seankyer opened the pull request:
Add alert notification service (ANS) to Bluetooth subsystem and accompanying sample.

Based on spec described in https://www.bluetooth.com/specifications/specs/alert-notification-service-1-0/ with bitfields taken from https://btprodspecificationrefs.blob.core.windows.net/gatt-specification-supplement/GATT_Specification_Supplement.pdf.

To interact with this sample, download a BLE app like ADI Attach or nRF connect. Once you subscribe to the unread alert and new alert notification characteristics, you need to send a "write" to the Alert Notification Control Point characteristic to enable the categories you would like to be notified on.

In the basic example you would send: [0,0] and [1,0] to the control point and once you have subscribed to notifications you will then start receiving the messages. Further commands are explained in https://btprodspecificationrefs.blob.core.windows.net/gatt-specification-supplement/GATT_Specification_Supplement.pdf section 3.12.

It is the responsibility of the app to interpret whatever a 'new notification' or 'unread notification' is and use the APIs to update them accordingly. The service must save the most recent value for each enabled category in case the control point sends the 'immediate' request, which relays the most up to date values for the specified category.

seankyer committed:
bluetooth: ANS: Add Alert Notification Service

Add alert notification service (ANS) to Bluetooth
subsystem and accompanying sample.

Signed-off-by: Sean Kyer <sean.actor@gmail.com>

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,96 @@
+ */
+#define BT_ANS_MAX_TEXT_STR_SIZE 18
+
+/**
+ * @brief ANS Category ID Type
+ */
+typedef uint8_t bt_ans_cat_t;
+
+/**
+ * @brief ANS Category ID Enum
```
If we have the enum, why do we need the typedef? Suggest to change to a named enum instead

seankyer replied:
Originally, it was so that I could use the enum directly in the construction of the packet (since the function signature took uint8_t). But I've taken your suggestion and I just cast the enum in the function that uses it now.

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,96 @@
+ * Enumeration for whether the category is supported.
+ */
+enum {
+	bt_ans_cat_simple_alert,
+	bt_ans_cat_email,
+	bt_ans_cat_news,
+	bt_ans_cat_call,
+	bt_ans_cat_missed_call,
+	bt_ans_cat_sms_mms,
+	bt_ans_cat_voice_mail,
+	bt_ans_cat_schedule,
+	bt_ans_cat_high_pri_alert,
+	bt_ans_cat_instant_message,
+	/* Marker for number of categories */
+	bt_ans_cat_num,
+	/* 10–15 reserved */
```
I don't know if we actually have any rules about this, but most (all?) enums in Zephyr are in caps, like `BT_ANS_CAT_SIMPLE_ALERT` instead of `bt_ans_cat_simple_alert`

Thalley replied:
Yeah, shouldn't be too difficult to verify that values between `enum {` and `};` are caps

https://docs.zephyrproject.org/latest/contribute/style/code.html states that we follow the Linux style guide, and they do actually state (https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl)

```
12) Macros, Enums and RTL

Names of macros defining constants and labels in enums are capitalized.
```

So we do have a rule for it, but seemingly it isn't checked.

Thalley replied:
Just FYI, created https://github.com/zephyrproject-rtos/zephyr/issues/95679 to fix this :)

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,96 @@
+};
+
+/**
+ * @brief Send a new alert to central
```
Suggest to change `central` to `remote devices` or something similar. 

The reason for this is that
1) The service may exist on a central device, and the peripheral may be the one who subscribed to the alerts
2) The service probably need to notify all connected clients

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,96 @@
+ */
+int bt_ans_notify_new_alert(bt_ans_cat_t category, uint8_t num_new, const char *text, int text_len);
+
+/**
+ * @brief Set the total unread count for a given category
+ *
+ * The unread count is transmitted to the central if notifications are enabled. Each category will
+ * save the latest call to this function in case an immediate replay is requested via the ANS
+ * control point.
+ *
+ * @param category The bt_ans_cat_t category the unread count is for
+ * @param unread Total number of unread alerts
+ *
+ * @return 0 on success
+ * @return negative error codes on failure
+ */
+int bt_ans_set_unread_count(bt_ans_cat_t category, uint8_t unread);
```
How does this work with multiple connected clients? The ANS spec isn't very clear on that IMO.

seankyer replied:
I am actually unclear myself if this supports multiple connected devices. The spec says that there can only be one instance of the service running. This function sets the unread count globally.

Thalley replied:
Right, but the unread count should ideally be set per remote device, right? If you add 2 new alerts, and one client reads 1 and the other read 2, then both will have `unread=0` after that, right?

Thalley commented on include/zephyr/bluetooth/att.h:
```diff
@@ -86,6 +86,8 @@ extern "C" {
  *
  * Defined by the Supplement to the Bluetooth Core Specification (CSS), v11, Part B, Section 1.2.
  */
+/** Command not supported */
+#define BT_ATT_ERR_CMD_NOT_SUP			0xa0
```
This is a service specific error code, and should be defined in the ans.h :)

Thalley commented on samples/bluetooth/peripheral_ans/src/main.c:
```diff
@@ -0,0 +1,82 @@
+	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, sizeof(CONFIG_BT_DEVICE_NAME) - 1),
+};
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		printk("Connection failed, err 0x%02x %s\n", err, bt_hci_err_to_str(err));
+	} else {
+		printk("Connected\n");
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	printk("Disconnected, reason 0x%02x %s\n", reason, bt_hci_err_to_str(reason));
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
```
Consider adding support for restarting advertising using the `recycled` callback

Thalley commented on samples/bluetooth/peripheral_ans/src/main.c:
```diff
@@ -0,0 +1,82 @@
+	printk("Sample - Bluetooth Peripheral ANS\n");
+
+	err = bt_enable(NULL);
+	if (err) {
```
Out coding guidelines (https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) rule 85 mandates that controlling statements shall be essentially boolean


```suggestion
	if (err != 0) {
```

Thalley replied:
Yeah, I've been trying to get that added for a while, but no luck yet :)

Thalley reviewed (changes requested):
Thanks for your PR - It's overall good quality, and most of my comments are related some potential missing features or Zephyr specific formatting or coding guidelines

seankyer commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,390 @@
+	     "At least one ANS New Alert category must be enabled in Kconfig");
+
+/* Statically configured via Kconfig */
+const static uint16_t alert_cat_id_bit_mask = BT_ANS_NALRT_CAT_MASK;
+const static uint16_t unread_cat_id_bit_mask = BT_ANS_UNALRT_CAT_MASK;
```
Actually, since we will configure this at runtime and not statically, I will just remove const.

seankyer commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,390 @@
+static struct unread_alert_status unread_alerts[bt_ans_cat_num];
+
+/* Initialize to 0, it is control point responsibility to enable once connected */
+static uint16_t alert_cat_enabled;
+static uint16_t unread_cat_enabled;
```
Added suffix "_map"

seankyer commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,390 @@
+{
+	/* Check if the category is supported */
+	if ((unread_cat_id_bit_mask & (1U << category)) == 0) {
+		LOG_ERR("Category %d unsupported", category);
```
Maybe an appropriate place to use WRN?

Thalley replied:
LOG_WRN could be used, yeah, but my opinion / rule of thumb is that if we return an error code directly to the call, LOG_DBG is what makes the most sense

seankyer committed:
doc: release: Add Bluetooth ANS service to 4.4 release notes

Add Alert Notification Service (ANS) and enabling Kconfig
to 4.4 release notes.

Signed-off-by: Sean Kyer <sean.actor@gmail.com>

seankyer commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,390 @@
+							    &new_alerts[category],
+							    sizeof(struct new_alert));
+					k_mutex_unlock(&new_alert_mutex);
+					if (rc) {
```
I have:
1. Added a check on the `rc` for the notify function to ignore the `-ENOTCONN` error code.
2. Added an assert to the top of `ans.c` to ensure max num of connections cannot exceed the number of buffers
3. I assume that assert would be in the notif function itself?
For all other error codes, we get an error and return the `rc`.

seankyer commented:
Thanks for the review. I replied to a few comments and applied your suggestions.

I also refactored the transmit functions since they got too nested.

seankyer commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @param category The bt_ans_cat_t category the notification is for
+ * @param num_new Number of new alerts since last alert
+ * @param text Text brief of alert
+ * @param text_len String length of text param. If greater than MAX_TEXT_STR_SIZE, text is truncated
```
And that the size is in terms of bytes

Thalley replied:
I think we should make it clear that the length here refer to the number of octets, and not the number of characters (for UTF8 those may differ significantly)

seankyer replied:
Added it to the API doc

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,96 @@
+ * @return 0 on success
+ * @return negative error codes on failure
+ */
+int bt_ans_notify_new_alert(bt_ans_cat_t category, uint8_t num_new, const char *text, int text_len);
```
In some services we let the service handle the truncation. Given that ANS (seemingly) use UTF8, it may be easier to do this for the users, rather than expect users to correctly use `utf8_trunc`

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,514 @@
+	strncpy(new_alerts[cat_in].text_string, text, size);
+	new_alerts[cat_in].text_string[size] = '\0';
+	ret = k_mutex_unlock(&new_alert_mutex);
+	if (ret != 0) {
+		LOG_ERR("Unable to unlock mutex (err: %d)", ret);
+		return ret;
+	}
```
Consider replacing this with

```suggestion
	__ASSERT(ret == 0, "Unable to unlock mutex (err: %d)", ret);
```

as if this fails to return, there is a fatal error in the code

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,514 @@
+int bt_ans_set_new_alert_support_category(enum bt_ans_category category, bool enable)
+{
+	uint8_t cat_in = (uint8_t)category;
+	bool has_conn = false;
+
+	bt_conn_foreach(BT_CONN_TYPE_ALL, ans_conn_check_cb, &has_conn);
+
+	if (has_conn) {
+		/* Cannot change support while connection exists */
+		return -EBUSY;
+	}
```
Consider moving this to a `static bool` function so that it can be reused by `bt_ans_set_unread_support_category`

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,514 @@
+		return -EAGAIN;
+	}
+
+	rc = bt_gatt_notify_uuid(NULL, BT_UUID_GATT_NALRT, ans_svc.attrs, &new_alerts[cat],
+				 sizeof(struct new_alert));
+
+	ret = k_mutex_unlock(&new_alert_mutex);
+	if (ret != 0) {
+		LOG_ERR("Unable to unlock mutex (err: %d)", ret);
+		return ret;
+	}
+
+	/* If the client is not connected, that is fine */
+	if (rc != 0 && rc != -ENOTCONN) {
```
`rc` from `bt_gatt_notify_uuid` gets overriden by `k_mutex_unlock`

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,514 @@
+	}
+
+	/* Otherwise send just the requested category (if enabled) */
+	if (unread_cat_enabled_map & BIT(cat_in)) {
```
```suggestion
	if ((unread_cat_enabled_map & BIT(cat_in)) != 0) {
```

"essential boolean" again :D

seankyer replied:
agh! thanks.

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,514 @@
+
+	int size = MIN(BT_ANS_MAX_TEXT_STR_SIZE - 1, text_len);
+
+	strncpy(new_alerts[cat_in].text_string, text, size);
+	new_alerts[cat_in].text_string[size] = '\0';
```
We should use `utf8_trunc` to truncate strings if they can be UTF8 encoded, otherwise the truncation here may remove part of a UTF8 character

Thalley replied:
This also means that `CONFIG_BT_ANS` needs `depends on UTF8`

seankyer replied:
okay added all that

Thalley reviewed (changes requested):
A handful of comments again :) 

I don't know the service or profile well enough to determine how well the API suits the use cases of the profile, but overall it looks pretty good as a service implementation

There are still some uncertainty on how to handle multiple connections. Perhaps the notify functions should take a `struct bt_conn *` as a parameter and then the application can decide whether to provide a specific conn or `NULL` to notify a single connection or all

Thalley commented on subsys/bluetooth/services/Kconfig.ans:
```diff
@@ -0,0 +1,84 @@
+module-str = Alert Notification Service (ANS)
+source "subsys/logging/Kconfig.template.log_config"
+
+menu "New Alert Categories"
+
+config BT_ANS_NALRT_CAT_SIMPLE_ALERT
+	bool "Support Simple Alert"
+
+config BT_ANS_NALRT_CAT_EMAIL
+	bool "Support Email"
+
+config BT_ANS_NALRT_CAT_NEWS
+	bool "Support News"
+
+config BT_ANS_NALRT_CAT_CALL
+	bool "Support Call"
+
+config BT_ANS_NALRT_CAT_MISSED_CALL
+	bool "Support Missed Call"
+
+config BT_ANS_NALRT_CAT_SMS_MMS
+	bool "Support SMS/MMS"
+
+config BT_ANS_NALRT_CAT_VOICE_MAIL
+	bool "Support Voice Mail"
+
+config BT_ANS_NALRT_CAT_SCHEDULE
+	bool "Support Schedule"
+
+config BT_ANS_NALRT_CAT_HIGH_PRI_ALERT
+	bool "Support High Prioritized Alert"
+
+config BT_ANS_NALRT_CAT_INSTANT_MESSAGE
+	bool "Support Instant Message"
+
+endmenu
```
Hmm, yeah, good point. One solution could be to have the service be dynamically registered, but that seems like a poor solution to this simple issue. 

It's not really an issue that they are here, but it is somewhat confusing to set the new alert as well as supported alerts in 2 places (both runtime and compile time). Not a blocker :)

seankyer commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,119 @@
+ * @return 0 on success
+ * @return negative error codes on failure
+ */
+int bt_ans_set_new_alert_support_category(enum bt_ans_category category, bool enable);
```
Yeah that is simpler, okay i updated it.

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,135 @@
+ *
+ * Enumeration for whether the category is supported.
+ */
+enum bt_ans_category {
+	BT_ANS_CAT_SIMPLE_ALERT,    /**< Simple alerts (general notifications). */
```
I'm not aware of any official guidelines on this, but I have a personal preference to use the same prefix for the enum and the values, i.e. either rename as


```suggestion
enum bt_ans_category {
	BT_ANS_CATEGORY_SIMPLE_ALERT,    /**< Simple alerts (general notifications). */
```

or

```suggestion
enum bt_ans_cat {
	BT_ANS_CAT_SIMPLE_ALERT,    /**< Simple alerts (general notifications). */
```

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,135 @@
+ * will save the latest call to this function in case an immediate replay is requested via the ANS
+ * control point.
+ *
+ * @note This function waits on a Mutex with K_FOREVER to ensure atomic updates to notification
```
```suggestion
 * @note This function waits on a Mutex with @ref K_FOREVER to ensure atomic updates to notification
```

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,135 @@
+ * @param category The category the notification is for
+ * @param num_new Number of new alerts since last alert
+ * @param text Text brief of alert
+ * @param text_len UTF8 string length of text param in octets. If greater than MAX_TEXT_STR_SIZE,
```
```suggestion
 * @param text_len UTF8 string length of text param in octets. If greater than @ref BT_ANS_MAX_TEXT_STR_SIZE,
```

seankyer replied:
just going to go with this: https://github.com/zephyrproject-rtos/zephyr/pull/95578#discussion_r2382750297

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,135 @@
+ * @param conn The connection object to send the alert to
+ * @param category The category the notification is for
+ * @param num_new Number of new alerts since last alert
+ * @param text Text brief of alert
+ * @param text_len UTF8 string length of text param in octets. If greater than MAX_TEXT_STR_SIZE,
+ * text is truncated
+ *
+ * @return 0 on success
+ * @return negative error codes on failure
+ */
+int bt_ans_notify_new_alert(struct bt_conn *conn, enum bt_ans_category category, uint8_t num_new,
+			    const char *text, int text_len);
```
w.r.t. having both `const char *text` and `text_len`, how do you feel about adding a requirement that `text` is NULL terminated (similar to `strxxx` functions from the stdlib)? That way we can omit the additional size/length field

seankyer replied:
Okay, lets do it that way, I agree.

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,496 @@
+struct new_alert {
+	uint8_t category_id;
+	uint8_t num_new_alerts;
+	char text_string[BT_ANS_MAX_TEXT_STR_SIZE];
```
```suggestion
	char text_string[BT_ANS_MAX_TEXT_STR_SIZE + 1];
```

Don't we need to 1 additional octet here for the NULL terminator?

seankyer replied:
I don't think so since the spec talks about this string in terms of octets. So I would assume the NULL terminator is part of the 'max size' already.

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,496 @@
+
+LOG_MODULE_REGISTER(bt_ans, CONFIG_BT_ANS_LOG_LEVEL);
+
+/* Build assert to ensure notifications cannot fail by not having enough buffers */
+BUILD_ASSERT(CONFIG_BT_MAX_CONN <= CONFIG_BT_ATT_TX_COUNT,
+	     "Max Connections must be less than or equal to allocated buffers");
```
This does not guarantee that it won't fail. 

If the device has `CONFIG_BT_MAX_CONN=4` and `CONFIG_BT_ATT_TX_COUNT=8` and there are other services that each have 3 outstanding ATT buffers, then we only have 2 remaining. 

It's a rare case, and unlikely to happen depending on what other services exists on the device, but the only robust solution is a retry mechanism. It is, however, not something that we have in all services, so I guess it's OK to continue with this, but the comment needs to be removed/redone as it's claiming something that's incorrect :)

seankyer replied:
Right, in a real system there may be multiple services in-use at a time. 

Okay, I will add more details to the comment. This assertion really only catches the lower bound at build time, but doesn't guarantee the problem you describe wont occur at runtime.

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,496 @@
+static uint16_t alert_cat_enabled_map;
+static uint16_t unread_cat_enabled_map;
+
+/* Notification variables configured by CCC */
+static bool new_alert_notify;
+static bool unread_alert_notify;
```
This won't work for `CONFIG_BT_MAX_CONN > 1`. 

Imagine Device A subscribing to alert status, i.e. `unread_alert_notify = (value == BT_GATT_CCC_NOTIFY);` sets `unread_alert_notify` to `true`. 
Then Device B does the same, and `unread_alert_notify` is still set to `true`. 
Then Device A removes its subscription, which sets `unread_alert_notify` to `false`, even when Device B is still subscribed.

You can use `bt_gatt_is_subscribed` to check this instead

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,496 @@
+
+	/* If the client is not connected, that is fine */
+	if (rc != 0 && rc != -ENOTCONN) {
+		LOG_ERR("Error notifying New Alert category %d rc: %d", cat, rc);
```
If `conn` is not subscribed, then this returns an error too, but it's not a fatal error, so `LOG_ERR` should be used. Additionally, `notify_new_alert_category` is called from `transmit_new_alert` and the return value is propagated back to the send which has 

```c
		rc = transmit_new_alert(conn, command.category);
		if (rc != 0) {
			LOG_ERR("Error transmitting New Alert during control point request! "
				"(rc=%d)",
				rc);
			return BT_GATT_ERR(BT_ANS_ERR_CMD_NOT_SUP);
		}
```

So if we fail to notify the new alert, we log that twice as a fatal/uncoverable error, and then we also reject the write command if the connection isn't subscribed. That doesn't seem like the correct/intended behavior, is it? 

Additionally, if we receive a `BT_ANS_CTRL_NOTIFY_NEW_ALERT_IMMEDIATE`, don't we need to send a notification to all connected devices, and not just the one that sent the control point operation?

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,496 @@
+	/* Special case: send all categories */
+	if (cat_in == BT_ANS_SEND_ALL_CATEGORY) {
+		for (int i = 0; i < BT_ANS_CAT_NUM; i++) {
+			if ((alert_sup_cat_bit_mask & BIT(i)) && (alert_cat_enabled_map & BIT(i))) {
```
```suggestion
			if (((alert_sup_cat_bit_mask & BIT(i)) != 0) && ((alert_cat_enabled_map & BIT(i)) != 0)) {
```

seankyer replied:
oh man.. it's such muscle memory. I really wish we had that compliance check 😅

Thalley replied:
Yep :)

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,496 @@
+	/* Truncate UTF8 string if too large */
+	int size = MIN(BT_ANS_MAX_TEXT_STR_SIZE - 1, text_len);
+
+	strncpy(new_alerts[cat_in].text_string, text, size);
+	new_alerts[cat_in].text_string[size] = '\0';
+	utf8_trunc(new_alerts[cat_in].text_string);
```
Did you consider using `utf8_lcpy` which correctly copies and truncates a UTF8 string, and ensures that it's NULL terminated? 

That assumes that `text` is NULL terminated, but so does `strncpy`

Thalley commented on subsys/bluetooth/services/Kconfig.ans:
```diff
@@ -0,0 +1,86 @@
+
+menu "New Alert Categories"
+
+config BT_ANS_NALRT_CAT_SIMPLE_ALERT
+	bool "Support Simple Alert"
+
+config BT_ANS_NALRT_CAT_EMAIL
+	bool "Support Email"
+
+config BT_ANS_NALRT_CAT_NEWS
+	bool "Support News"
+
+config BT_ANS_NALRT_CAT_CALL
+	bool "Support Call"
+
+config BT_ANS_NALRT_CAT_MISSED_CALL
+	bool "Support Missed Call"
+
+config BT_ANS_NALRT_CAT_SMS_MMS
+	bool "Support SMS/MMS"
+
+config BT_ANS_NALRT_CAT_VOICE_MAIL
+	bool "Support Voice Mail"
+
+config BT_ANS_NALRT_CAT_SCHEDULE
+	bool "Support Schedule"
+
+config BT_ANS_NALRT_CAT_HIGH_PRI_ALERT
+	bool "Support High Prioritized Alert"
+
+config BT_ANS_NALRT_CAT_INSTANT_MESSAGE
+	bool "Support Instant Message"
```
I guess we can assume that people will realize that `select EXPERIMENTAL` is inherited given the `if BT_ANS`, right? 

Not sure if we need to `select EXPERIMENTAL` on all of these

Thalley reviewed (changes requested):
Some things that I think needs to be fixed :)

Thalley commented on samples/bluetooth/peripheral_ans/src/main.c:
```diff
@@ -0,0 +1,121 @@
+		}
+		k_sleep(K_SECONDS(1));
+
+		ret = bt_ans_set_unread_count(NULL, BT_ANS_CAT_SIMPLE_ALERT, ++num_unread);
```
`num_unread` can overflow here, can't it?

seankyer replied:
For this and https://github.com/zephyrproject-rtos/zephyr/pull/95578#discussion_r2387080560 it will overflow and loop back around to 0 (essentially restarting the sample). I thought this was fine for the sample. I'll add a comment to the static declaration of the variables however stating this behavior.

Thalley replied:
Yeah, it's fine to overflow, but ideally we explicitly state this to avoid any confusion :)

Thalley commented on samples/bluetooth/peripheral_ans/src/main.c:
```diff
@@ -0,0 +1,121 @@
+		static const char test_msg[] = "Test Alert!";
+		static const char high_pri_msg[] = "Prio Alert!";
+
+		num_new++;
```
`num_new` can overflow here, can't it?

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,491 @@
+	__ASSERT(ret == 0, "Unable to lock mutex (err: %d)", ret);
+	new_alerts[cat_in].num_new_alerts = num_new;
+
+	int size = MIN(BT_ANS_MAX_TEXT_STR_SIZE - 1, strlen(text));
+
+	utf8_lcpy(new_alerts[cat_in].text_string, text, size);
```
The last argument to `utf8_lcpy` is the destination buffer size, so we should be able to just do


```suggestion
	utf8_lcpy(new_alerts[cat_in].text_string, text, sizeof(new_alerts[cat_in].text_string));
```

Thalley reviewed (changes requested):
A few minor comments, but otherwise LGTM. 

Please also have a look at the issues reported by Sonar

Thalley commented on subsys/bluetooth/services/ans.c:
```diff
@@ -0,0 +1,491 @@
+	__ASSERT(ret == 0, "Unable to lock mutex (err: %d)", ret);
+	new_alerts[cat_in].num_new_alerts = num_new;
+
+	int size = MIN(BT_ANS_MAX_TEXT_STR_SIZE - 1, strlen(text));
+
+	utf8_lcpy(new_alerts[cat_in].text_string, text, size);
```
The 3rd argument should be the size of the destination buffer as per https://github.com/zephyrproject-rtos/zephyr/pull/95578#discussion_r2387103809

Thalley reviewed (dismissed):
I think this is good for now - The spec itself is too unclear about how to handle multiple connections, so that part is still somewhat unclear in the implementation as well. 

It would have been nice with some testing of this service, but that would also require a client implementation

kartben commented on samples/bluetooth/peripheral_ans/README.rst:
```diff
@@ -0,0 +1,29 @@
+   :name: Peripheral ANS
+   :relevant-api: bluetooth
+
+   Sample application for Alert Notification Service (ANS).
```
```suggestion
   Send notification using Alert Notification Service (ANS).
```

kartben reviewed (dismissed):
+1 for docs / sample, thanks!

jhedberg commented:
@seankyer yes, that's unfortunately the case. It still needs a (re-)review from @Thalley.

kartben commented:
sorry @jhedberg @seankyer - I didn't see the mention! Yeah, I typically add the "future" docs around that time. Guess it would make sense to make it part of the check list for release managers so that we don't forget in the future :)
https://github.com/zephyrproject-rtos/zephyr/pull/98537

Thalley commented on include/zephyr/bluetooth/services/ans.h:
```diff
@@ -0,0 +1,134 @@
+	BT_ANS_CAT_HIGH_PRI_ALERT,  /**< High-priority alerts. */
+	BT_ANS_CAT_INSTANT_MESSAGE, /**< Instant messaging alerts. */
+
+	BT_ANS_CAT_NUM, /**< Marker for the number of categories. */
```
We should consider omitting this particular value from the documentation

```suggestion
	/** @cond INTERNAL_HIDDEN */
	BT_ANS_CAT_NUM, /**< Marker for the number of categories. */
	/** @endcond */
```

Thalley commented on samples/bluetooth/peripheral_ans/README.rst:
```diff
@@ -0,0 +1,26 @@
+************
+
+* A board with Bluetooth LE support
+* Smartphone with BLE app (ADI Attach, nRF Connect, etc.) or dedicated BLE sniffer
```
```suggestion
* Smartphone with Bluetooth LE app (ADI Attach, nRF Connect, etc.) or dedicated Bluetooth LE sniffer
```

IMO we should avoid using the term "BLE", and instead rely on the official term "Bluetooth LE"

Thalley reviewed (changes requested):
Two minor comments, otherwise LGTM

seankyer commented on subsys/bluetooth/services/Kconfig.ans:
```diff
@@ -0,0 +1,86 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig BT_ANS
+	bool "GATT Alert Notification Service"
```
oops, I misinterpreted what you meant and instead moved the `select EXPERIMENTAL` to the end of the Kconfig. Okay, I added that to the description.
